### ISSUE TITLE
[Idea] So I've a bunch of suggestions about things I'd like improved.

### ISSUE THREAD

**Issue by Slayer418** [tajemniktv/TajsGraph#28]
I'm gonna share the post I made on their QA forum, will be simplier : https://questions.satisfactorygame.com/post/671be17bddb9d97e072086fd

**Comment by tajemniktv** [comment]
Contact shadows *should* be disabled by default (or by the user, I might've missed that) as they are... uhm, well shitty as you noticed yourself in your post XD
VSMs are (or at least should) be a replacement for Contact Shadows. Tho VSMs can't seem to create shadows for hypertubes or power pole cables (shit i forgor, gotta make an issue for that ig)

But overall I didn't touch a lot on the shadows except for allowing to use VSMs (and earlier disabling cascaded completely). I will look into them more, once I get the base rework more complete.

Disabling far shadow culling is such a performance hog I think it's not worth it, although I did (will) add a placeholder (switch) for enabling/disabling it. 

The day and night transition got even it's own issue (#18) cuz I feel like it's somehow terrible... For now I'm only looking into how things are rendered, so it's more of a "do it later" issue tho, but I find the sun leaking into buildings (mostly during transitions) extremely annoying... And as you noticed, it's also often like "a switch". 

>Second issue, a weird foliage/grass type that obey the related commands differently. I can't figure how to properly increase its render distance but toying with grass.CullDistanceScale and r.ViewDistanceScale have some mitigated impact on it.

That's also a really good find and I think it's directly related to #21. I thought it's TajsGraph breaking stuff, but turns out it's vanilla being vanilla. Not really sure what might be causing it, but I'm also on the lookout for possible causes, tho as I said, rn I'm kinda focued on reworking the base, which in theory from outwside might look really similiar to how things have been (tho im hoping for a pretty performance boost), but damn inside it's completely different lol 

I will rename the issue so it better represents what I should look into (but what wasn't alr in previous issues)

**Comment by tajemniktv** [comment]
I forgot to add - Thanks for reporting this both on QA site and here, really appreciate the detailed explanations along with screenshots - that really makes things a bit easier!
Cheers:D

**Comment by Slayer418** [comment]
It's a pleasure my guy. I'm kinda obsessed with tweaking UE games in general, I've been doing it for for years now. Trying to get a game to look great, with no immersion breaking stuff occuring (stuff pop'ing in, fading in/out, etc) and whitout impacting perf too much, it's quite a challenge. Satisfactory tho is quite very well optimized in comparison with ARK Survival Ascended for example.

As for Contact Shadows, I kinda wish we could tweak them, not disable them. Like increase their quality.

**Comment by Sinner-Man77** [comment]
Hi Tajemnik and Slayer,

I have an explanation for the shadows of hypertubes. It’s not a Virtual Shadows problem; it’s a LOD problem. Hypertubes, like machines for instance, behave like foliage with foliage.LODDistanceScale, and the shadows depend on the LOD. At LOD 0, only the circles appear; at LOD 1, no shadows appear; with a higher LOD, the tube casts a full shadow.

So when you are far away (depending on foliage.LODDistanceScale), the shadow is full, then as you get closer, it disappears, and the circles appear. This can be easily tested with foliage.ForceLOD (default -1).

More generally, regarding the shadow issue, the best compromise seems to me (my opinion is subject to change…) to play with VS and DFS. To avoid the horrible bubble effect of the very limited r.Shadow.DistanceScale, I set it to 0, which in return creates very ugly tree shadows (you can then adjust r.DFShadowQuality, r.DFShadow.TwoSidedMeshDistanceBiasScale, and r.DistanceFields.SurfaceBiasExpand). You can favor acceptable tree shadows by setting r.DistanceFields.SurfaceBiasExpand to 0 or 0.001 (different contrast from the foliage), for example, at the expense of the quality of other shadows and accepting other issues, or choose a higher value (which may be better if you don't care too much about tree shadows).

Other remarks:
- To avoid missing and popping VS (at least with the space elevator), I also use r.SkeletalMeshLODBias=-5, r.StaticMeshLODDistanceScale=0.01, and r.MipMapLodBias=-2 (this one I don’t remember why);
- r.AOGlobalDistanceField.NumClipmaps=8 helps avoid a problem of shadow size change with the DFS;
- r.Shadow.RadiusThreshold helps avoid missing and popping VS, but is very resource-intensive (a low value like 0.01, 0.02, or 0.03 is however inevitable when only using VS);
- unrelated: the waterfalls look really bad, I haven’t managed to improve their quality, so I use fx.EnableNiagaraSpriteRendering=0.
You can test this:
r.Shadow.Virtual.Enable 1
r.ViewDistanceScale 5
foliage.LODDistanceScale 5
r.Shadow.DistanceScale 0
r.AOGlobalDistanceField.NumClipmaps 8
r.DFShadow.TwoSidedMeshDistanceBiasScale 1
r.DistanceFields.SurfaceBiasExpand 0 (or higher value)
r.DFShadowQuality 2 (or 3?)
r.SkeletalMeshLODBias -5
r.StaticMeshLODDistanceScale 0.01
r.MipMapLodBias -2
r.Shadow.RadiusThreshold 0.02

I hope I haven’t said too much nonsense...

Sinner-Man

**Comment by Slayer418** [comment]
Yeah I found out that r.Shadow.RadiusThreshold is the culprit for my too perf hungry shadows. I now have my sg.ShadowQuality set to 3 and r.Shadow.RadiusThreshold forced to 0.1 to not have my FPS tank too much. It causes some (very few) StaticMesh shadows to appear/disappear at a sadly not too far distance tho (the lower r.Shadow.RadiusThreshold, the further away those shadows stay) but it's fine, I turned ContactShadow back on to mitigate that.
